### PR TITLE
[NT-878, NT-873] Add Experiment and Read More is a full-width button

### PR DESCRIPTION
**Pull request by cdolm92**
# 📲 What

Added the new Optimizely experiment to test project campaign CTA.

# 🤔 Why

This work is part of our experimentation to test whether we can increase conversion from the campaign details page.

# 🛠 How

We have a `projectCampaignExperiment()` that gives us the `ProjectCampaignExperimentType` based on the experiment variation the user is in. The type can be `.experimental` or `control`.

Also created a `ProjectCampaignButtonStyleType` to determine the style of the button based on `ProjectCampaignExperimentType`.  

# 👀 See

Trello, screenshots, external resources?

| Experiment 🐛 | Control 🦋 |
| --- | --- |
| Simulator Screen Shot - iPhone 11 - 2020-02-19 at 13 13 41| Simulator Screen Shot - iPhone 11 2 - 2020-02-19 at 13 14 47|

# ✅ Acceptance criteria

You can whitelist your device-id on Optimizely by looking for the property `session_device_distinct_id` which is fired with most events.
- [ ] Whitelisting your device-id into the `control` group: navigate to any project, you should see "Read more about the campaign ->" as the CTA copy.
- [ ] Whitelisting your device-id into the `variant-1` group: navigate to any project, you should see the experimental "Read more about the campaign" full-width button
- [ ] Whitelisting your device-id into the `variant-2` group: navigate to any project, you should see the experimental "Read more about the campaign" full-width button

### CODE REVIEW

File: Library/Styles/ButtonStyles.swift [hunk 1]
```diff
@@ -132,3 +132,23 @@ public let shareButtonStyle =
   <> UIButton.lens.image(for: .normal) .~ image(named: "icon--share")
   <> UIButton.lens.tintColor .~ .ksr_soft_black
   <> UIButton.lens.accessibilityLabel %~ { _ in Strings.dashboard_accessibility_label_share_project() }
+
+// MARK: - Read More Campaign Button
+
+public let controlReadMoreButtonStyle = baseButtonStyle
```
Comment (justinswart) [inline Library/Styles/ButtonStyles.swift]: I think we could separate the fact that the style is control/experimental from the naming of it, so I would call this something like `readMoreButtonTextStyle`.

File: Library/Styles/ButtonStyles.swift [hunk 2]
```diff
@@ -132,3 +132,23 @@ public let shareButtonStyle =
...
+  <> UIButton.lens.contentEdgeInsets .~ .init(
+    top: Styles.grid(3) - 1,
+    left: 0,
+    bottom: Styles.grid(4) - 1,
+    right: 0
+  )
+  <> UIButton.lens.backgroundColor .~ .white
+
+// experimental campaign button
+public let experimentalReadMoreButtonStyle = greyButtonStyle
```
Comment (justinswart) [inline Library/Styles/ButtonStyles.swift]: This could just be `readMoreButtonRoundedGreyStyle` maybe?

File: Library/Styles/ButtonStyles.swift [hunk 3]
```diff
@@ -132,3 +132,23 @@ public let shareButtonStyle =
...
   <> UIButton.lens.tintColor .~ .ksr_soft_black
   <> UIButton.lens.accessibilityLabel %~ { _ in Strings.dashboard_accessibility_label_share_project() }
+
+// MARK: - Read More Campaign Button
+
+public let controlReadMoreButtonStyle = baseButtonStyle
+  <> UIButton.lens.titleColor(for: .normal) .~ .ksr_soft_black
+  <> UIButton.lens.titleColor(for: .highlighted) .~ .ksr_text_dark_grey_500
+  <> UIButton.lens.titleLabel.font .~ .ksr_headline(size: 15)
+  <> UIButton.lens.title(for: .normal) %~ { _ in Strings.Read_more_about_the_campaign_arrow() }
```
Comment (justinswart) [inline Library/Styles/ButtonStyles.swift]: Could we set the title text in the VC and separate it from the style?

File: Library/Styles/ProjectStyles.swift [hunk 4]
```diff
@@ -3,6 +3,50 @@ import Prelude
 import Prelude_UIKit
 import UIKit
 
+public enum ProjectCampaignExperimentType {
```
Comment (justinswart) [inline Library/Styles/ProjectStyles.swift]: I don't think we need this interstitial type. The button style can just be a function in the view model:

```swift
public func projectCampaignButtonStyleForVariant(
  _ variant: OptimizelyExperiment.Variant
) -> ProjectCampaignButtonStyleType {
  switch variant {
  case .control:
    return .controlReadMoreButton
  case .variant1, .variant2:
    return .experimentalReadMoreButton
  }
}
```
And then the signals in `ProjectPamphletMainCellViewModel` can just be:
```swift
let projectCampaignExperimentType = self.awakeFromNibProperty.signal
  .map { projectCampaignExperiment() }
  .skipNil()
self.readMoreButtonStyle = projectCampaignExperimentType.map(projectCampaignButtonStyleForVariant)
self.spacerViewHidden = projectCampaignExperimentType.map { $0 != .control }
self.blurbAndReadMoreStackViewSpacing = projectCampaignExperimentType.map { $0 == .control ? 0 : 4 }
  .map(Styles.grid)
```

File: Library/Styles/ProjectStyles.swift [hunk 5]
```diff
@@ -3,6 +3,50 @@ import Prelude
...
+    switch self {
+    case .control:
+      return Styles.grid(0)
+    case .experimental:
+      return Styles.grid(4)
+    }
+  }
+}
+
+public enum ProjectCampaignButtonStyleType: Equatable {
```
Comment (justinswart) [inline Library/Styles/ProjectStyles.swift]: This type helps us with the VM output for testing 👍

File: Library/ViewModels/ProjectPamphletMainCellViewModel.swift [hunk 6]
```diff
@@ -417,3 +436,30 @@ private func progressColor(forProject project: Project) -> UIColor {
...
 }
+
+private func projectCampaignExperiment() -> ProjectCampaignExperimentType {
+  let userAttributes = optimizelyUserAttributes(
+    with: AppEnvironment.current.currentUser,
+    project: nil,
+    refTag: nil
+  )
+
+  let optimizelyVariant = AppEnvironment.current.optimizelyClient?
```
Comment (justinswart) [inline Library/ViewModels/ProjectPamphletMainCellViewModel.swift]: If we remove `ProjectCampaignExperimentType` then we can just return this directly as `OptimizelyExperiment.Variant?`.

File: Library/ViewModels/ProjectPamphletMainCellViewModel.swift [hunk 7]
```diff
@@ -417,3 +436,30 @@ private func progressColor(forProject project: Project) -> UIColor {
     return .ksr_green_700
   }
 }
+
+private func projectCampaignExperiment() -> ProjectCampaignExperimentType {
+  let userAttributes = optimizelyUserAttributes(
+    with: AppEnvironment.current.currentUser,
+    project: nil,
```
Comment (justinswart) [inline Library/ViewModels/ProjectPamphletMainCellViewModel.swift]: Note to self to add `project` and `refTag` here once this merges with #1072.

**Comment by nativeksr** [comment]
### SwiftFormat found issues:

| File | Rules |
| ---- | ----- |
| Library/ViewModels/ProjectPamphletMainCellViewModel.swift | braces, consecutiveSpaces, indent |


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Library/ViewModels/ProjectPamphletMainCellViewModel.swift [hunk 8]
```diff
@@ -29,6 +29,9 @@ public protocol ProjectPamphletMainCellViewModelOutputs {
   /// Emits a string to use for the backers title label.
   var backersTitleLabelText: Signal<String, Never> { get }
 
+  /// Emits the spacing of the blurb and reward stgack view.
```
Comment (justinswart) [inline Library/ViewModels/ProjectPamphletMainCellViewModel.swift]: Typo in this comment.

Comment (cdolm92) [reply Library/ViewModels/ProjectPamphletMainCellViewModel.swift]: I corrected it

**Pull request merged**